### PR TITLE
Poll for notification permission changes

### PR DESCRIPTION
**Pull request by emawby**
# Description
## One Line Summary
This PR adds polling for notification permission changes so that we can detect changes that happen from outside of the SDK in the current session.

## Details
Spawn a new thread where we check notification permission every second. If the permission has changed we fire the permission observer.
This means that we need to track the permission value to know if it has changed. 

This is done using an `_enabled` boolean that gets set when using OneSignal's prompting or when the polling detects a change. 

When the app loses focus we change the polling interval to 1 day to have the thread sleep until the app is brought into focus. 

### Motivation
Currently if an app prompts for permissions without using OneSignal or if the user changes permissions without closing the app (notification center long press on notification), we don't detect those permission changes until the next session begins.

### Scope
notification permission prompting

# Testing
## Unit testing
Added a unit testing file for the PermissionController. Currently has 3 tests around the new polling behavior.

## Manual testing
Tested by prompting directly through Android native APIs to test that the SDK detects the change.
Tested that backgrounding that properly sleeps the thread to ensure we aren't polling while the app is not in focus.

# Affected code checklist
   - [x] Notifications
      - [ ] Display
      - [ ] Open
      - [ ] Push Processing
      - [ ] Confirm Deliveries
   - [ ] Outcomes
   - [ ] Sessions
   - [ ] In-App Messaging
   - [ ] REST API requests
   - [ ] Public API changes

# Checklist
## Overview
   - [x] I have filled out all **REQUIRED** sections above
   - [x] PR does one thing
     - If it is hard to explain how any codes changes are related to each other then it most likely needs to be more than one PR
   - [x] Any Public API changes are explained in the PR details and conform to existing APIs

## Testing
   - [x] I have included test coverage for these changes, or explained why they are not needed
   - [x] All automated tests pass, or I explained why that is not possible
   - [x] I have personally tested this on my device, or explained why that is not possible

## Final pass
   - [x] Code is as readable as possible.
      - Simplify with less code, followed by splitting up code into well named functions and variables, followed by adding comments to the code.
   - [x] I have reviewed this PR myself, ensuring it meets each checklist item
      - WIP (Work In Progress) is ok, but explain what is still in progress and what you would like feedback on. Start the PR title with "WIP" to indicate this.

- - -
This change is [Reviewable](https://reviewable.io/reviews/OneSignal/OneSignal-Android-SDK/2112)

### CODE REVIEW

File: OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt
Comment (jkasten2) [inline OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt:75]: Seems like the AndroidX library' `areNotificationsEnabled() takes care of all Android versions:
https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:core/core/src/main/java/androidx/core/app/NotificationManagerCompat.java;l=364-389?q=notificationmanagercompat

File: OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt
Comment (jkasten2) [inline OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt:145]: This might be a good time to put these 3 lines into its own function.

File: OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt
Comment (jkasten2) [inline OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt:73]: Not sure if `_application.appContext` is set at this point, `notificationsEnabled()` will always return `false` if it is not.

File: OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt
Comment (jkasten2) [inline OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt:87]: Note, might want to use a waiter here with a timeout to account for waking when the app is foregrounded, and changing the wait time to something very large when the app is backgrounded (to effectively pause the polling).

File: OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt
Comment (emawby) [reply OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt:75]: It has always been set and returns the correct value in my testing, but that is not to say that it couldn't happen. If it isn't then we will just fire the handler unnecessarily which isn't the worst thing. We could also save the value in sharedPreferences to reduce the likely hood of this occurring.

Comment (jkasten2) [reply OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt:75]: I wasn't actually sure off hand so wanted to make sure we know if this was the case / tested. Since it is working I think it's fine to set it like this.
* However I can see this being fragile, but caching it has it own problem, as it could be out of date.

File: OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt
Comment (emawby) [reply OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt]: I updated this to use a waiter. The in focus interval is 1 second and the out of focus is currently 30 minutes. I am not sure what a reasonable value is for the out of focus polling. Do we even want to continue polling while in the background?

File: OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt
Comment (jkasten2) [inline OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt:62]: We don't send a value so we can simply this a bit use `Waiter` instead.

File: OneSignalSDK/onesignal/core/src/main/java/com/onesignal/core/internal/config/ConfigModel.kt
Comment (jkasten2) [inline OneSignalSDK/onesignal/core/src/main/java/com/onesignal/core/internal/config/ConfigModel.kt:214]: Let's use a high value of 1 day if we are going down this timmer route.

File: OneSignalSDK/onesignal/core/src/main/java/com/onesignal/core/internal/config/ConfigModel.kt
Comment (jkasten2) [inline OneSignalSDK/onesignal/core/src/main/java/com/onesignal/core/internal/config/ConfigModel.kt:211]: Add comment noting that we don't really want to poll in the background, but using a very large value to essentially stop polling in the background.

**Review by jkasten2 — CHANGES_REQUESTED** [review]
Comments are for if we keep this change the polling rate for the background to a value so high it essentially stops polling.

Still for stopping the polling loop and starting a new one when foregrounding if you want to go that route though, as the intent in the code would be more clear.

File: Examples/OneSignalDemo/app/src/main/java/com/onesignal/sdktest/model/MainActivityViewModel.java
Comment (jkasten2) [inline Examples/OneSignalDemo/app/src/main/java/com/onesignal/sdktest/model/MainActivityViewModel.java:814]: Remove commented out code.

**Review by jkasten2 — COMMENTED** [review]
The implementation itself looks good to me now.

This PR is still noted as WiP, so I'll wait to approve until it's considered complete.

File: OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt
Comment (jkasten2) [inline OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt:80]: We should start with the background time here to be safe. I believe the onFcous event fires right away if the app is already in focus so it will be updated right away if the app happens to really be in the foreground at this point.

**Review by jkasten2 — CHANGES_REQUESTED** [review]
After looking through the test I have a concern about the default `pollingWaitInterval` value, left more details on the line of code.

File: OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt
Comment (emawby) [reply OneSignalSDK/onesignal/notifications/src/main/java/com/onesignal/notifications/internal/permissions/impl/NotificationPermissionController.kt]: Good catch! Yes it does fire immediately and wakes so it starts polling correctly

**Review by jkasten2 — CHANGES_REQUESTED** [review]
New "NotificationPermissionController permission polling fires permission changed event" test is failing on CI from the last commit. I believe the test needs to be updated to account for the new polling rate default.

**Review by jkasten2 — APPROVED** [review]
Test passes locally now for me, so it's most likely good now, but let's wait until CI finishes before merging to be sure.

**Pull request merged**
